plugins/modules/files/archive.py

```diff
@@ -378,7 +376,15 @@ def add_targets(self):
             )
 
     def compare_with_original(self):
-        self.changed |= self.original_size != self.destination_size()
+        if self.original_checksums is None:
+            self.changed |= self.original_size != self.destination_size()
+        else:
+            self.changed |= self.original_checksums != self.destination_checksums()
```

Comment:
Also, regardless of the outcome of the previous comment wrt this method, I believe the OR-ing of `changed` is misplaced. I know this was already there before this PR, but I think this method should not be concerned with `changed` and it should only return the result of the comparison. Then, down there in `main()`, lines 647 and 657, the OR-assignment to `self.changed` should be explicit.

Reply:
Makes sense since this doesn't alter state. Applied with the last commit.